addons/survey/models/survey_user.py

```diff
@@ -250,92 +248,94 @@ def write(self, vals):
+    def save_lines(self, user_input, question, answer, comment=None):
         """ Save answers to questions, depending on question type
 
             If an answer already exists for question and user_input_id, it will be
             overwritten (or deleted for 'choice' questions) (in order to maintain data consistency).
         """
-        vals = {
-            'user_input_id': user_input_id,
-            'question_id': question.id,
-            'survey_id': question.survey_id.id,
-            'skipped': False,
-        }
         old_answers = self.search([
-            ('user_input_id', '=', user_input_id),
+            ('user_input_id', '=', user_input.id),
             ('survey_id', '=', question.survey_id.id),
             ('question_id', '=', question.id)
         ])
 
         if question.question_type in ['char_box', 'text_box', 'numerical_box', 'date', 'datetime']:
-            self._save_line_simple_answer(vals, old_answers, question, answer)
+            self._save_line_simple_answer(user_input, question, old_answers, answer)
         elif question.question_type in ['simple_choice', 'multiple_choice']:
-            self._save_line_choice(vals, old_answers, question, answer, comment)
+            self._save_line_choice(user_input, question, old_answers, answer, comment)
         elif question.question_type == 'matrix':
-            self._save_line_matrix(vals, old_answers, answer, comment)
+            self._save_line_matrix(user_input, question, old_answers, answer, comment)
         else:
             raise AttributeError(question.question_type + ": This type of question has no saving function")
 
     @api.model
-    def _save_line_simple_answer(self, vals, old_answers, question, answer):
-        answer_type = question.question_type
-
-        vals.update(self._get_save_line_values(answer, answer_type))
+    def _save_line_simple_answer(self, user_input, question, old_answers, answer):
+        vals = self._get_line_answer_values(user_input, question, answer, question.question_type)
         if old_answers:
             old_answers.write(vals)
+            return old_answers
         else:
-            self.create(vals)
-        return True
+            return self.create(vals)
 
     @api.model
-    def _save_line_choice(self, vals, old_answers, question, answers, comment):
+    def _save_line_choice(self, user_input, question, old_answers, answers, comment):
         if not (isinstance(answers, list)):
             answers = [answers]
 
-        vals_list = []
         if question.question_type == 'simple_choice':
             if not (question.comment_count_as_answer and question.comments_allowed and comment):
-                for answer in answers:
-                    vals.update(self._get_save_line_values(answer, 'suggestion'))
-                    vals_list.append(vals.copy())
+                vals_list = [self._get_line_answer_values(user_input, question, answer, 'suggestion') for answer in answers]
```

Comment:
I remember there was some trickyness to this, but @dbeguin will have more insights.

Reply:
Previously there was a shared dict as basis for all values (base values then an update of line specific values), meaning a copy was necessary to avoid updating the same dict at each iteration.

Now a new dict is generated at each save_line call, for each answer (in _get_line_answer_values and _get_line_comment_values).